pkg/app_watcher/appwatcher.go

```diff
@@ -0,0 +1,162 @@
+package app_watcher
+
+import (
+	"context"
+	appclientset "github.com/argoproj/argo-cd/v2/pkg/client/clientset/versioned"
+	"github.com/rs/zerolog/log"
+	"github.com/zapier/kubechecks/pkg/config"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"strings"
+	"time"
+
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	applisters "github.com/argoproj/argo-cd/v2/pkg/client/listers/application/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ApplicationWatcher is the controller that watches ArgoCD Application resources via the Kubernetes API
+type ApplicationWatcher struct {
+	cfg                  *config.ServerConfig
+	applicationClientset appclientset.Interface
+	appInformer          cache.SharedIndexInformer
+	appCache             cache.InformerSynced
+	appLister            applisters.ApplicationLister
+}
+
+// NewApplicationWatcher creates new instance of ApplicationWatcher.
+func NewApplicationWatcher(cfg *config.ServerConfig) (*ApplicationWatcher, error) {
+	// this assumes kubechecks is running inside the cluster
+	kubeCfg, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		log.Fatal().Msgf("Error building kubeconfig: %s", err.Error())
+	}
+
+	appClient := appclientset.NewForConfigOrDie(kubeCfg)
+
+	ctrl := ApplicationWatcher{
+		cfg:                  cfg,
+		applicationClientset: appClient,
+	}
+
+	appInformer, appLister := ctrl.newApplicationInformerAndLister()
+
+	ctrl.appInformer = appInformer
+	ctrl.appLister = appLister
+
+	return &ctrl, nil
+}
+
+// Run starts the Application CRD controller.
+func (ctrl *ApplicationWatcher) Run(ctx context.Context, processors int) {
+	log.Info().Msg("starting Application Controller")
+
+	defer runtime.HandleCrash()
+
+	go ctrl.appInformer.Run(ctx.Done())
+
+	if !cache.WaitForCacheSync(ctx.Done(), ctrl.appInformer.HasSynced) {
+		log.Error().Msg("Timed out waiting for caches to sync")
+		return
+	}
+
+	<-ctx.Done()
+}
+
+// onAdd is the function executed when the informer notifies the
+// presence of a new Application in the namespace
+func (ctrl *ApplicationWatcher) onApplicationAdded(obj interface{}) {
+	if !canProcessApp(obj) {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.Error().Err(err).Msg("appwatcher: could not get key for added application")
+	}
+	log.Trace().Str("key", key).Msg("appwatcher: onApplicationAdded")
+	ctrl.cfg.VcsToArgoMap.AddApp(obj.(*appv1.Application))
+}
+
+func (ctrl *ApplicationWatcher) onApplicationUpdated(old, new interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(new)
+	if err != nil {
+		log.Warn().Err(err).Msg("appwatcher: could not get key for updated application")
+	}
+	// TODO
+	// have any of the Source repoURLs changed?
+	log.Trace().Str("key", key).Msg("appwatcher: onApplicationUpdated")
```

Comment:
at this point we'll have to remove the old app (in all the various maps), then add the new one back in

Reply:
yeah that's a smart way to handle it instead of doing an upsert